pkg/sql/opt/indexrec/index_recommendation_set.go

```diff
@@ -0,0 +1,151 @@
+// Copyright 2021 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package indexrec
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/cockroachdb/cockroach/pkg/sql/opt"
+	"github.com/cockroachdb/cockroach/pkg/sql/opt/cat"
+	"github.com/cockroachdb/cockroach/pkg/sql/opt/memo"
+)
+
+// FindIndexRecommendationSet finds index candidates that are used in an
+// expression to determine a statement's index recommendation set.
+func FindIndexRecommendationSet(expr opt.Expr, md *opt.Metadata) string {
+	recommendationSet := indexRecommendationSet{md: md}
+	recommendationSet.init()
+	recommendationSet.createIndexRecommendations(expr)
+	return recommendationSet.getStringOutput()
+}
+
+// indexRecommendationSet stores the hypothetical indexes that are used in a
+// statement's optimal plan (in usedIndexes), as well as the statement's
+// metadata.
+type indexRecommendationSet struct {
+	md          *opt.Metadata
+	usedIndexes map[cat.Table][]int
+}
+
+// init initializes an indexRecommendationSet by allocating memory for it.
+func (irs *indexRecommendationSet) init() {
+	numTables := len(irs.md.AllTables())
+	irs.usedIndexes = make(map[cat.Table][]int, numTables)
+}
+
+// createIndexRecommendations recursively walks an expression tree to find
+// hypothetical indexes that are used in it.
+func (irs *indexRecommendationSet) createIndexRecommendations(expr opt.Expr) {
+	switch expr := expr.(type) {
+	case *memo.ScanExpr:
+		irs.addIndexToRecommendationSet(expr.Index, expr.Table)
+	case *memo.LookupJoinExpr:
+		irs.addIndexToRecommendationSet(expr.Index, expr.Table)
+	case *memo.InvertedJoinExpr:
+		irs.addIndexToRecommendationSet(expr.Index, expr.Table)
+	}
+	for i, n := 0, expr.ChildCount(); i < n; i++ {
+		irs.createIndexRecommendations(expr.Child(i))
+	}
+}
+
+// addIndexToRecommendationSet adds an index to the indexes map if it does not
+// exist already in the map and in the table.
+func (irs *indexRecommendationSet) addIndexToRecommendationSet(
+	indexOrd cat.IndexOrdinal, tabID opt.TableID,
+) {
+	switch hypTable := irs.md.TableMeta(tabID).Table.(type) {
+	case *hypotheticalTable:
+		// Do not add real table indexes (non-hypothetical indexes).
+		if indexOrd < hypTable.Table.IndexCount() {
+			return
+		}
+		for _, existingIndex := range irs.usedIndexes[hypTable] {
+			if existingIndex == indexOrd {
```

Comment:
I didn't use a set here as suggested in a prototype review comment because there were issues using it in conjunction with a map.